public/app/features/explore/RichHistory/RichHistory.tsx

```diff
@@ -0,0 +1,219 @@
+    store.set(RICH_HISTORY_SETTING_KEYS.activeDatasourceOnly, activeDatasourceOnly);
+  };
+
+  onSelectDatasourceFilters = (value: SelectableValue[]) => {
+    this.setState({ datasourceFilters: value });
+  };
+
+  onSelectTab = (item: SelectableValue<Tabs>) => {
+    this.setState({ activeTab: item.value! });
+  };
+
+  onChangeSortOrder = (sortOrder: SortOrder) => this.setState({ sortOrder });
+
+  render() {
+    const {
+      datasourceFilters,
+      sortOrder,
+      activeTab,
+      starredTabAsFirstTab,
+      activeDatasourceOnly,
+      retentionPeriod,
+    } = this.state;
+    const { theme, richHistory, onChangeRichHistoryProperty, activeDatasourceInstance } = this.props;
+    const styles = getStyles(theme);
+
+    const QueriesTab = {
+      label: 'Query History',
+      value: Tabs.RichHistory,
+      content: (
+        <RichHistoryContent
+          onlyStarred={false}
+          queries={richHistory}
+          sortOrder={sortOrder}
+          datasourceFilters={datasourceFilters}
+          activeDatasourceOnly={activeDatasourceOnly}
+          activeDatasourceInstance={activeDatasourceInstance}
+          retentionPeriod={retentionPeriod}
+          onChangeSortOrder={this.onChangeSortOrder}
+          onChangeRichHistoryProperty={onChangeRichHistoryProperty}
+          onSelectDatasourceFilters={this.onSelectDatasourceFilters}
+        />
+      ),
+      icon: 'fa fa-history',
+    };
+
+    const StarredTab = {
+      label: 'Starred',
+      value: Tabs.Starred,
+      content: (
+        <RichHistoryContent
+          onlyStarred={true}
+          queries={richHistory}
+          sortOrder={sortOrder}
+          datasourceFilters={datasourceFilters}
+          activeDatasourceOnly={activeDatasourceOnly}
+          activeDatasourceInstance={activeDatasourceInstance}
+          retentionPeriod={retentionPeriod}
+          onChangeSortOrder={this.onChangeSortOrder}
+          onChangeRichHistoryProperty={onChangeRichHistoryProperty}
+          onSelectDatasourceFilters={this.onSelectDatasourceFilters}
+        />
+      ),
+      icon: 'fa fa-star',
+    };
+
+    const SettingsTab = {
+      label: 'Settings',
+      value: Tabs.Settings,
+      content: (
+        <RichHistorySettings
+          retentionPeriod={this.state.retentionPeriod}
+          starredTabAsFirstTab={this.state.starredTabAsFirstTab}
+          activeDatasourceOnly={this.state.activeDatasourceOnly}
+          onChangeRetentionPeriod={this.onChangeRetentionPeriod}
+          toggleStarredTabAsFirstTab={this.toggleStarredTabAsFirstTab}
+          toggleactiveDatasourceOnly={this.toggleactiveDatasourceOnly}
+        />
+      ),
+      icon: 'gicon gicon-preferences',
+    };
+
+    let tabs = [];
+    tabs.push(QueriesTab);
+    tabs.push(SettingsTab);
+
+    /* If user selects to have starredTabAsFirstTab === true, move
+     * StarredTab to first position, otherwise second
+     */
+    tabs.splice(starredTabAsFirstTab ? 0 : 1, 0, StarredTab);
```

Comment:
nitpick: would probably be clearer if we just do
```
tabs = starredTabAsFirstTab ? [StarredTab, QueriesTab, SettingsTab] : [QueriesTab, StarredTab, SettingsTab]
```

Reply:
💙beautiful and clean solution.